test/infrastructure/kubemark/controllers/kubemarkmachine_controller.go

```diff
@@ -0,0 +1,451 @@
+	if err := r.Get(context.TODO(), types.NamespacedName{
+		Name:      machine.Spec.Bootstrap.ConfigRef.Name,
+		Namespace: machine.Spec.Bootstrap.ConfigRef.Namespace,
+	}, &kubeadmConfig); err != nil {
+		logger.Error(err, "error getting bootstrap config")
+		return ctrl.Result{}, err
+	}
+
+	cfg, err := RetrieveValidatedConfigInfo(kubeadmConfig.Spec.JoinConfiguration)
+	if err != nil {
+		logger.Error(err, "error validating token")
+		return ctrl.Result{}, err
+	}
+
+	clusterinfo := cfg.Clusters[""]
+	cfg = CreateWithToken(
+		clusterinfo.Server,
+		DefaultClusterName,
+		TokenUser,
+		clusterinfo.CertificateAuthorityData,
+		kubeadmConfig.Spec.JoinConfiguration.Discovery.BootstrapToken.Token,
+	)
+	certificateStore := &capkcert.MemoryStore{}
+
+	newClientFn := func(current *tls.Certificate) (clientset.Interface, error) {
+		client, err := clientset.NewForConfig(restConfig)
+		if err != nil {
+			logger.Error(err, "error creating clientset")
+			return nil, err
+		}
+		return client, nil
+	}
+	mgr, err := certificate.NewManager(&certificate.Config{
+		BootstrapCertificatePEM: cfg.AuthInfos[TokenUser].ClientCertificateData,
+		BootstrapKeyPEM:         cfg.AuthInfos[TokenUser].ClientKeyData,
+		CertificateStore:        certificateStore,
+		Template: &x509.CertificateRequest{
+			Subject: pkix.Name{
+				CommonName:   fmt.Sprintf("system:node:%s", kubemarkMachine.Name),
+				Organization: []string{"system:nodes"},
+			},
+		},
+		Usages: []certificates.KeyUsage{
+			certificates.UsageDigitalSignature,
+			certificates.UsageKeyEncipherment,
+			certificates.UsageClientAuth,
+		},
+		ClientsetFn: newClientFn,
+	})
+	if err != nil {
+		logger.Error(err, "error creating cert manager")
+		return ctrl.Result{}, err
+	}
+
+	mgr.Start()
+
+	for {
+		_, err := certificateStore.Current()
+		if err != nil {
+			if _, ok := err.(*certificate.NoCertKeyError); !ok {
+				logger.Error(err, "err fetching certificate")
+				return ctrl.Result{}, err
+			}
+
+			time.Sleep(time.Second)
+
+			continue
+		}
+
+		break
+	}
+	mgr.Stop()
+
+	kubeconfig, err := generateCertificateKubeconfig(restConfig, "/kubeconfig/cert.pem")
+	if err != nil {
+		logger.Error(err, "err generating certificate kubeconfig")
+		return ctrl.Result{}, err
+	}
+
+	stackedCert := bytes.Buffer{}
+	if err := pem.Encode(&stackedCert, &pem.Block{Type: cert.CertificateBlockType, Bytes: certificateStore.Certificate.Leaf.Raw}); err != nil {
+		logger.Error(err, "err encoding certificate")
+		return ctrl.Result{}, err
+	}
+	keyBytes, err := keyutil.MarshalPrivateKeyToPEM(certificateStore.Certificate.PrivateKey)
+	if err != nil {
+		logger.Error(err, "err encoding key")
+		return ctrl.Result{}, err
+	}
+	stackedCert.Write(keyBytes)
+
+	configMap := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      kubemarkMachine.Name,
+			Namespace: kubemarkMachine.Namespace,
+		},
+		Data: map[string]string{
+			"kubeconfig": string(kubeconfig),
+			"cert.pem":   stackedCert.String(),
```

Comment:
Is this configmap something that is required by kubemark, or our implementation detail? Given the contents here it would be nice to use a Secret vs a ConfigMap.

Reply:
Looking below, it looks like it's our implementation detail, and the same volume mounting should be available with a Secret